modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/ImageItem.ts

```diff
@@ -0,0 +1,88 @@
+import { AlloyComponent, Disabling, ItemTypes, Toggling, Tooltipping } from '@ephox/alloy';
+import { Menu } from '@ephox/bridge';
+import { Fun, Merger, Optional } from '@ephox/katamari';
+import { Class, Html, Insert, SelectorFind, SugarElement } from '@ephox/sugar';
+
+import { UiFactoryBackstageProviders } from 'tinymce/themes/silver/backstage/Backstage';
+
+import * as ItemClasses from '../ItemClasses';
+import ItemResponse from '../ItemResponse';
+import { renderItemStructure } from '../structure/ItemStructure';
+import { buildData, renderCommonItem } from './CommonMenuItem';
+
+const renderImgItem = (
+  spec: Menu.ImageMenuItem,
+  onItemValueHandler: (itemValue: string) => void,
+  isSelected: boolean,
+  itemResponse: ItemResponse,
+  providersBackstage: UiFactoryBackstageProviders
+): ItemTypes.ItemSpec => {
+  const getApi = (component: AlloyComponent): Menu.ToggleMenuItemInstanceApi => ({
+    setActive: (state) => {
+      Toggling.set(component, state);
+    },
+    isActive: () => Toggling.isOn(component),
+    isEnabled: () => !Disabling.isDisabled(component),
+    setEnabled: (state) => Disabling.set(component, !state)
+  });
+
+  const structure = renderItemStructure({
+    presets: 'img',
+    textContent: Optional.none(),
+    htmlContent: Optional.none(),
+    ariaLabel: spec.tooltip,
+    iconContent: Optional.some(spec.url),
+    labelContent: spec.label,
+    shortcutContent: Optional.none(),
+    checkMark: Optional.none(),
+    caret: Optional.none(),
+    value: spec.value
+  }, providersBackstage, true);
+
+  const optTooltipping = spec.tooltip
+    .map((t) => Tooltipping.config(
+      providersBackstage.tooltips.getConfig({
+        tooltipText: providersBackstage.translate(t)
+      })
+    ));
+
+  return Merger.deepMerge(
+    renderCommonItem({
+      context: spec.context,
+      data: buildData(spec),
+      enabled: spec.enabled,
+      getApi,
+      onAction: (api) => {
+        onItemValueHandler(spec.value);
+        api.setActive(true);
+      },
+      onSetup: (api) => {
+        api.setActive(isSelected);
+        return Fun.noop;
+      },
+      triggersSubmenu: false,
+      itemBehaviours: [
+        ...optTooltipping.toArray()
+      ]
+    }, structure, itemResponse, providersBackstage),
+    {
+      toggling: {
+        toggleClass: ItemClasses.tickedClass,
+        toggleOnExecute: false,
+        exclusive: true,
+        onToggled: (comp: AlloyComponent) => {
+          SelectorFind.descendant(comp.element, '.tox-collection__item-image').each((imgContainer) => {
+            const checkmarkIcon = providersBackstage.icons().checkmark;
+            const iconElement = SugarElement.fromTag('div');
+            Class.add(iconElement, 'tox-collection__item-image-check');
+            Html.set(iconElement, checkmarkIcon);
+            Insert.append(imgContainer, iconElement);
+          });
```

Comment:
Spocke was right in his earlier comment, this should be a `replacing` behaviour in the image component.

In addition to making assumptions about the image component structure, which will make it harder to change, this looks like it will add a new checkmark every time it's toggled?

Reply:
Not sure we should even use that, just make the image item a toggle menu item. Compare with ToggleMenuItem.
 So this becomes:
```
 toggling: {
        toggleClass: ItemClasses.tickedClass,
        toggleOnExecute: false,
        selected: spec.active,
        exclusive: true
      }
``` 
Then replace `checkMarks: Optional.none()` with Optional.some(renderCheckmark(providers.icons))`

And then in ItemStructure.ts you render the checkmark the same way one does it for the toggle menu item.

That way the toggles state gets updated on render and uses the same method as other toggle menu items.

Reply:
But the difference is that the checkmark then get rendered outside the image item. But I wonder if that could be done with CSS instead to place it on top.

Reply:
Did some more refactor on that branch you can put that checkmark anywhere in the structure just passed it into the ImageSpec.